src/marlin_pc/constraints.rs

```diff
@@ -0,0 +1,1978 @@
+use crate::{
+    data_structures::LabeledCommitment,
+    kzg10::{Proof, VerifierKey as KZG10VerifierKey},
+    marlin_pc::{
+        data_structures::{Commitment, VerifierKey},
+        MarlinKZG10, PreparedCommitment, PreparedVerifierKey,
+    },
+    pc_constraints::{EvaluationsVar, PCCheckVar, QuerySetVar},
+    vec, BTreeMap, BTreeSet, BatchLCProof, LinearCombinationVar, PrepareVar, String, ToString, Vec,
+};
+use ark_ec::{CycleEngine, PairingEngine};
+use ark_ff::{fields::Field, PrimeField, ToConstraintField};
+use ark_nonnative_field::{NonNativeFieldMulResultVar, NonNativeFieldVar};
+use ark_poly::UVPolynomial;
+use ark_r1cs_std::{
+    alloc::{AllocVar, AllocationMode},
+    bits::{boolean::Boolean, uint8::UInt8, ToBitsGadget},
+    eq::EqGadget,
+    fields::{fp::FpVar, FieldVar},
+    groups::CurveVar,
+    pairing::PairingVar,
+    select::CondSelectGadget,
+    R1CSVar, ToBytesGadget, ToConstraintFieldGadget,
+};
+use ark_relations::r1cs::{ConstraintSystemRef, Namespace, Result as R1CSResult, SynthesisError};
+use ark_std::ops::Div;
+use core::{borrow::Borrow, convert::TryInto, marker::PhantomData, ops::MulAssign};
+
+/// Var for the verification key of the Marlin-KZG10 polynomial commitment scheme.
+#[allow(clippy::type_complexity)]
+pub struct VerifierKeyVar<
+    CycleE: CycleEngine,
+    PG: PairingVar<CycleE::E2, <CycleE::E1 as PairingEngine>::Fr>,
+> where
+    <CycleE::E2 as PairingEngine>::G1Projective: MulAssign<<CycleE::E1 as PairingEngine>::Fq>,
+    <CycleE::E2 as PairingEngine>::G2Projective: MulAssign<<CycleE::E1 as PairingEngine>::Fq>,
+    <CycleE::E2 as PairingEngine>::G1Affine:
+        ToConstraintField<<<CycleE::E1 as PairingEngine>::Fr as Field>::BasePrimeField>,
+    <CycleE::E2 as PairingEngine>::G2Affine:
+        ToConstraintField<<<CycleE::E1 as PairingEngine>::Fr as Field>::BasePrimeField>,
```

Comment:
@npwardberkeley or @weikengchen do you mind checking if this is still necessary on the latest nightly/upcoming stable? I believe https://github.com/rust-lang/rust/pull/73905 might help with some of these unnecessary bounds (for `MulAssign`, for example)